streams/src/test/java/org/apache/kafka/test/InternalProcessorContextMock.java

```diff
@@ -0,0 +1,376 @@
+
+        private String applicationId;
+        private TaskId taskId;
+        private Serde<?> keySerde;
+        private Serde<?> valueSerde;
+        private File stateDir;
+        private StreamsMetricsImpl metrics;
+        private ProcessorRecordContext recordContext;
+        private StreamsConfig config;
+        private ProcessorNode processorNode;
+        private ThreadCache cache;
+
+        private final Map<String, StateStore> stateStoreMap;
+        private final ToInternal toInternal;
+
+        public Builder() {
+            this(new MockProcessorContext());
+        }
+
+        public Builder(final MockProcessorContext processorContext) {
+            mock = mock(InternalProcessorContext.class);
+            this.processorContext = processorContext;
+
+            stateStoreMap = new HashMap<>();
+            toInternal = new ToInternal();
+            setRecordContext(new ProcessorRecordContext(-1L, -1L, -1, null, new RecordHeaders()));
+            appConfigs(null);
+
+            applicationId = processorContext.applicationId();
+            taskId = processorContext.taskId();
+            keySerde = processorContext.keySerde();
+            valueSerde = processorContext.valueSerde();
+            stateDir = processorContext.stateDir();
+            metrics = (StreamsMetricsImpl) processorContext.metrics();
+        }
+
+
+        public Builder cache(final ThreadCache cache) {
+            this.cache = cache;
+            return this;
+        }
+
+        public Builder appConfigs(Map<String, Object> config) {
+            if (config == null) {
+                config = new HashMap<>();
+            }
+            if (!config.containsKey(StreamsConfig.APPLICATION_ID_CONFIG)) {
+                config.put(StreamsConfig.APPLICATION_ID_CONFIG, "");
+            }
+            if (!config.containsKey(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG)) {
+                config.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "");
+            }
+            this.config = new QuietStreamsConfig(config);
+            return this;
+        }
+
+        public Builder metrics(final StreamsMetricsImpl metrics) {
+            this.metrics = metrics;
+            return this;
+        }
+
+        public Builder stateDir(final File stateDir) {
+            this.stateDir = stateDir;
+            return this;
+        }
+
+        public Builder valueSerde(final Serde<?> valueSerde) {
+            this.valueSerde = valueSerde;
+            return this;
+        }
+
+        public Builder keySerde(final Serde<?> keySerde) {
+            this.keySerde = keySerde;
+            return this;
+        }
+
+        public Builder taskId(final TaskId taskId) {
+            this.taskId = taskId;
+            return this;
+        }
+
+        public Builder applicationId(final String applicationId) {
+            this.applicationId = applicationId;
+            return this;
+        }
+
+        public InternalProcessorContext build() {
```

Comment:
req: Please add a `build()` method that conditionally call `replay()` in the end. Something like `buildWithoutReplaying()`. `build()` should call `buildWithoutReplaying()` and then call `replay()`. The reason is that in a test additional behaviour could be added to the mock. The test author would call the `buildWithoutReplaying()`, add her behaviour to the mock and replay manually.

Reply:
Once `EasyMock` sets an expectation, we cannot change it. 

For example
```java
interface Math {
        int add(int a, int b);
        int sub(int a, int b);
}
```

```java
final Math math = EasyMock.mock(Math.class);
EasyMock.expect(math.add(1, 1)).andStubReturn(2);
EasyMock.expect(math.add(1, 1)).andStubReturn(3);
EasyMock.replay(math);
assertEquals(3, math.add(1, 1)); // AssertionFailedError: expected: <3> but was: <2>
```

We can expose a `get` method that returns the mock without setting any expectation, then the test author sets custom expectations on the mock, and then the test author calls the `build` method of the `builder` that sets default expectations.

```java
final var ipc = builder.get();
EasyMock.expect(ipc. /* ... */ ).andStubReturn(/* ... */)
builder.build();
```

Reply:
This works:
```
        final Math math = EasyMock.mock(Math.class);
        EasyMock.expect(math.add(1, 1)).andStubReturn(2);
        EasyMock.expect(math.add(1, 1)).andReturn(3);
        EasyMock.replay(math);
        assertEquals(3, math.add(1, 1));
```
This is also the use pattern I have in mind, because if one specifies additional behaviour, she most likely also wants that the calls are verified. Otherwise the mock could have been build with a different value for the return.

Reply:
Done.